q2_metadata/normalization/_norm_errors.py

```diff
@@ -0,0 +1,204 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2020, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import yaml
+
+
+class Error(Exception):
+    """Base class for exceptions encountered
+    in the rules checking."""
+
+    @classmethod
+    def generic_message(cls, rule, variable, value):
+        reformatted_value = '\t# %s' % yaml.dump(value).replace('\n', '\n\t# ')
+        return 'Wrong formatting for "%s" rule; variable %s:\n' \
+               '%s' % (rule, variable, reformatted_value)
+
+
+class RuleError(Error):
+    """Exception raised for error due to the
+    presence of a rule that's not recognized."""
+
+    def __init__(self, rule, variable, value):
+        self.rule = rule
+        self.variable = variable
+        self.value = value
+
+    def __str__(self):
+        message = self.generic_message(self.rule, self.variable, self.value)
+        return '%s -> rule not recognized' % message
+
+
+class ExpectedError(Error):
+    """Exception raised for errors in the
+    formatting of the "expected values" rules.
+    """
+    def __init__(self, variable, value):
+        self.rule = 'expected'
+        self.variable = variable
+        self.value = value
+        self.message = self.get_message()
+
+    def get_message(self):
+        if not isinstance(self.value, list):
+            return 'is not a list'
+        str_items = [x for x in self.value if isinstance(x, str)]
+        if len(str_items) != len(self.value):
+            return 'not all items are strings'
+
+    def __str__(self):
+        message = self.generic_message(self.rule, self.variable, self.value)
+        return '%s -> %s' % (message, self.message)
+
+
+class OntologyError(Error):
+    """Exception raised for errors in the
+    formatting of the "ontology values" rules.
+    """
+    def __init__(self, variable, value):
+        self.rule = 'ontology'
+        self.variable = variable
+        self.value = value
+        self.message = self.get_message()
+
+    def get_message(self):
+        if not isinstance(self.value, str):
+            return 'is not a string'
+        ontologies = ['Gazetteer ontology']
+        if self.value not in ontologies:
+            return 'none of %s' % ', '.join(ontologies)
```

Comment:
It is a little bit worrying that the term `Gazetteer ontology` is repeated here from the `check_ontology` function. Is this absolutely needed?

Reply:
ontologies must be one of a few allowed lists of terms (gazetteer, uberon. emp, and a few others to come about). How would you propose to check that passed ontology (i.e. string giving the name of the ontology) is one of the allowed ontologies otherwise?

Reply:
that should happen somewhere else, not within an exception...

Reply:
this is not an exception. 
this is part of the check_ontology() function, called by the check_rule(), which is verifying the proper formatting of the rule ontology (if present for the current variable).
This check_ontology() function return a boolean to check_rule(), that either collect the rule, or then only raises an exception, i.e. the exception is already somewhere else.

That said, if we are to not raise and end the qiime metadata normalize call, but rather collect all the expections in an overarching object in order to return to the user a more complete summary of all that may have gone wring during the formatting check (but also maybe during the matching of the rules content with the metadata content), we may think of doing this at another somewhere.

Reply:
you're right but if the exception fires here, it is not a re-check, just showing the allowed ones as a help to the user. Are you suggesting that the exception here reads or gets the list of allowed ontologies from somewhere else instead of having it hard-copied? For where them would be best? a file that's in some MANIFEST? I would say that it'd be an attribue of the exception class..

Reply:
e.g.
```
class OntologyError(Error):
    """Exception raised for errors in the
    formatting of the "ontology values" rules.
    """
    def __init__(self, variable, value):
        self.rule = 'ontology'
        self.variable = variable
        self.value = value
        self.message = self.get_message()
        self.ontologies = ['Gazetteer ontology']

    def get_message(self):
        if not isinstance(self.value, str):
            return 'is not a string'
        if self.value not in self.ontologies:
            return 'none of %s' % ', '.join(self.ontologies)

    def __str__(self):
        message = self.generic_message(self.rule, self.variable, self.value)
        return '%s -> %s' % (message, self.message)
```

Reply:
these user-defined errors are now removed as the error info are now collected during the checking (see function `check_rule()`). The term `Gazetteer ontology` as well as other allowed values are now colelcted once nd from one place that's not the code: the assets file `rules_allowed.yml`